Fix zarr loading

yonibrowning opened the pull request:
Starting this pull request knowing that it is going to need some work before merging.

My goal here is to refactor the registration pipeline to use aind-zarr-utils and remove (as much as possible) any numpy or orientation conversions. The existing version works, but it is *very* hard to re-use the pipeline-produced transforms after the fact. Unfortunately, this conversion will necessitate also updating any downstream code, making it a braking change. The good news is that this will make incorporating new templates/transforms much easier in the future!


List of changes (so far):
- aind-zarr-utils requires python >3.10; I have upgraded to 3.12, and made changes to the docker file so that all packages are compatible with this
- switched __read_zarr to aind_zarr_utils.zarr.zarr_to_ants
- registration is now done without resample (volume is not 25,25,25, but instead at automatically detected zarr level closest to 25,25,25)
- no rotations, transposes are applied after zarr image is loaded, with the exception of reorienting ants image for creating ccf-space zarrs. This step is necessary because the zarr does not include direction metadata
- updated plotting to handle anatomical information from ants without need for numpy conversion
- temporarily added command line/CodeOcean interface for testing. This includes a "flip" option for old data with incorrect metadata. Both of these should be removed prior to release.


I am still testing to make sure everything works as expected, once it does we can start to integrate.

Commit:
Configuring to work in app panel for testing

Commit:
Edited app-panel.json, metadata.yml

Commit:
Edited app-panel.json

Commit:
Deleted main.py; added main__.py; edited run

Commit:
Edited register.py

Commit:
Fiddled with Docker to make aind-zarr-utils install. Upgraded to python 3.12.4 in process.

Commit:
3.12 now builds properly.

Commit:
Works for single channel. Multiple channels needs testing.

Commit:
Docker file  updated to prevent numpy error in zarr writing

Commit:
Bug fixes and cleanup.

Commit:
Caught more should-be compound transforms.

Commit:
Bugfix

Commit:
added aind_registration_utils to docker.

Commit:
Runs to completion!

Commit:
Changed sidecar referencing.

Commit:
Documentation and typing updates

Commit:
Upped number of iterations, should ensure convergence.

Commit:
Added a few more debugging image saves.

Commit:
CHeckin- rigid not working, need to roll back

camilolaiton commented on code/aind_ccf_reg/preprocess.py:
```diff
@@ -239,6 +242,37 @@ def compute_mask(self, ants_img):
 
         return ants_img_mask
 
+    def _resample_to_spacing(self, img, new_spacing, interp_type=0):
```
Use typing hints

camilolaiton commented on code/aind_ccf_reg/preprocess.py:
```diff
@@ -239,6 +242,37 @@ def compute_mask(self, ants_img):
+        """
+        Resample ANTs image to a new spacing.
+
+        interp_type codes for ants.resample_image:
```
Use numpy docs style

camilolaiton commented on code/aind_ccf_reg/preprocess.py:
```diff
@@ -250,55 +284,221 @@ def compute_N4(
-            -- image {ants_img_n4}"
-        )
+        # Decide whether to do low-res estimation
+        use_lowres = target_spacing is not None
```
Could you please remind me why this is important?

camilolaiton commented on code/aind_ccf_reg/preprocess.py:
```diff
@@ -250,55 +284,221 @@ def compute_N4(
+    # ):
+    #     """N4 Bias Field Correction
+    #     https://antspy.readthedocs.io/en/latest/utils.html#ants.utils.bias_correction.n4_bias_field_correction
+    #     """
```
If this N4 code is dead, let's remove it

camilolaiton commented on code/aind_ccf_reg/register.py:
```diff
@@ -279,6 +280,11 @@ def register_to_template(
         )
+        
+        ants.image_write(ants_fixed,'/results/ants_fixed.nii.gz')
+        ants.image_write(ants_moving,'/results/ants_moving.nii.gz')
```
This would not work in pipeline mode. The output path should be given by parameter

camilolaiton replied:
Let's not assume absolute paths pointing to the results folder, in case someone external wants to use this package. In addition, this will not work in pipeline mode since pipelines do not have absolute paths defined (or did they change this?). Either way, we should use a parameter that points to the output path everywhere in the code.

camilolaiton commented on code/aind_ccf_reg/register.py:
```diff
@@ -346,9 +352,9 @@ def register_to_template(
         # The idea here being to ensure that the algorithm actually converges
         if self.args["reference_res"] == 25:
-            reg_iterations = [1,1,1,1]#[3000, 3000, 3000, 3000]
+            reg_iterations = [1,1,1,1]#[500, 100, 50, 0]
```
Why registering to template the affine iterations here are set to 1 in all pyramid levels?

camilolaiton commented on code/aind_ccf_reg/register.py:
```diff
@@ -364,8 +370,8 @@ def register_to_template(
             "mask_all_stages": True,
-            "syn_metric": "CC",
-            "syn_sampling": 2,
+            "syn_metric": "Mattes",
```
Why switching to mattes?

camilolaiton reviewed (changes requested):
A couple questions of the why behind specific changes. If you have a presentation with the results of why switching algorithms and iterations, that'd be awesome. (Also if you have the jacobbians)

Commit:
Changing so that registrations actually builds

Commit:
Affine runs now, too

galenlynch committed:
fix: keep the CCF parcellation in the zarr's axis order for precomputed

apply_transforms returns aligned_image in ants_img's geometry -- the
lightsheet's frame, and since zarr_to_ants preserves the zarr's axis order,
its (z, y, x) storage order. Reorienting it into the atlas frame straight
afterwards kept the anatomy correct for every affine-aware reader while
silently rolling the Neuroglancer overlay, because a precomputed volume
carries no affine: its info holds size and resolution and nothing else, so
the array order is the only thing tying the parcellation to the zarr it
parcellates.

That is why nothing errored. The bytes were correct under a wrong
description -- every file valid, CloudVolume content with it, Slicer showing
a perfect overlay from the sidecar nii.gz -- and the one fact neither format
records is the one that was wrong.

Also restore the C/F relabelling that has to follow. OME-Zarr describes its
axes C-order (z, y, x); precomputed describes the same bytes F-order
(x, y, z), so reversing the axis order translates between the two. It moves
no data -- a strided view, and a.tobytes('C') == a.T.tobytes('F') -- and the
spacing is reversed with it so the declaration matches the array. `.T`
rather than swapaxes(0, 2): reversing ALL axes is what the relabelling
means, the two coincide only in 3D, and CloudVolume's array is
[x, y, z, channel].

Verified against the captured 776259 output. Reconstructing the pre-reorient
array recovers IPR (480, 1120, 944), matching prep_zarr_img exactly; `.T`
then gives (944, 1120, 480) at (14400, 14400, 16000) -- the pipeline's own
info, and shape x 8 == the fused zarr's (7552, 8960, 3840). Deriving the
target independently from acquisition.json (z,y,x -> I,P,R, so precomputed
is RPI) produces a byte-identical array.

Note for reviewers: moved_ccf_anno_to_ls.nii.gz now lands in the lightsheet
frame (IPR, 480x1120x944) rather than the atlas frame. Anatomically
identical and still correct in Slicer -- it now shares an index grid with
the lightsheet volume -- but the shape differs from previous runs.

galenlynch committed:
fix(plots): regrid to canonical LPS with aind-anatomical-utils

canonical_lps_reference_enclosing built its reference grid by measuring the
image's world extent and dividing by its spacing. Those are indexed
differently -- the extent along world axes, the spacing in the image's own
index order -- so they only line up when the index axes already match the
world axes. Flips preserve that order and were fine; any permutation paired
the wrong spacing with the wrong extent and produced a wrong-shaped grid,
which the subsequent resample then distorted.

The volumes involved are axis-aligned, so canonicalising should be
shape-preserving. On the CCF annotation it was not: 944x1120x480 in, a
944x1009x534 reference out. Both the pre- and post-fix orientations of that
volume are permuted relative to LPS, so the QC PNGs have been geometrically
distorted throughout -- which is part of why a rolled Neuroglancer overlay
was never caught from them.

regrid_axis_aligned_ants does this properly, permuting and flipping to the
target orientation rather than constructing a grid by hand. It is invariant
to the input's index order and, for an axis-aligned volume, exact -- so it
also stops label volumes being blurred by the old interp_type="linear".

In plot_reg, `fixed` is regridded and the other two are resampled onto it,
so all three still share one grid for the overlay while the grid no longer
depends on `fixed`'s index order.

Verified: plot_antsimgs now emits byte-identical PNGs for the same volume
expressed in RAI/PIR/IPR/RPI/ASL/LPS index order; previously all six
differed. aind-anatomical-utils is already present via aind-zarr-utils
(<0.5,>=0.4), so no environment change.

pad_mm is now unused in both functions, and interpolation is unused in
plot_antsimgs. Signatures left alone -- no caller passes either -- but they
are worth removing.

flake8 on this file drops 22 -> 16 findings and interrogate 37.5% -> 50%;
both already failed repo-wide before this change.
